
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>A gentle introduction &#8212; py-dsyre 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Symbolic Regression" href="../theory/symbolic_regression.html" />
    <link rel="prev" title="Welcome to py-dsyre’s documentation!" href="../index.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="A-gentle-introduction">
<h1>A gentle introduction<a class="headerlink" href="#A-gentle-introduction" title="Permalink to this heading">¶</a></h1>
<section id="Preamble">
<h2>Preamble<a class="headerlink" href="#Preamble" title="Permalink to this heading">¶</a></h2>
<p>In this tutorial we will look into <em>py-dsyre</em> expression encoding. To explain it, it is perhaps best to start from a concrete example.</p>
<p>Consider the following mathematical formula <span class="math notranslate nohighlight">\(f(x,c) = \frac{x+\sin(cx)}{cx}\)</span> containing one independent variable <span class="math notranslate nohighlight">\(x\)</span> and one constant <span class="math notranslate nohighlight">\(c\)</span>.</p>
<p>A (possible) decomposition in simple, at most binary, operations is:</p>
<div class="math notranslate nohighlight">
\[\begin{split}\begin{array}{l}
    u_0 = x \\
    u_1 = c \\
    u_2 = u_0 \cdot u_1 \\
    u_3 = \sin(u_2) \\
    u_4 = u_3+u_0 \\
    u_5 = u_4 / u_2
\end{array}\end{split}\]</div>
<p>This sequence of decompositions is what <em>py-dsyre</em> encodes in a fixed length <strong>genotype</strong> made of integer triplets, each representing one of the <span class="math notranslate nohighlight">\(u_i\)</span>.</p>
<p>Assigning, in this case, to the single operations the ids: <span class="math notranslate nohighlight">\(+: 0, \cdot: 1, /: 2, \sin: 3\)</span> the above sequence of decompositions is uniquely determined by the triplets:</p>
<div class="math notranslate nohighlight">
\[\mathbf x = [\underbrace{1, 0, 1}_{u_2}, \underbrace{3, 2, 0}_{u_3}, \underbrace{0, 3, 0}_{u_4}, \underbrace{2, 4, 2}_{u_5}]\]</div>
<p>Note that the independent variable and parameters define the first <span class="math notranslate nohighlight">\(u_0=x\)</span> and <span class="math notranslate nohighlight">\(u_1=c\)</span>.</p>
<p>Any <strong>genotype</strong> is then easily expressed into the various <span class="math notranslate nohighlight">\(u_i\)</span>, which are ALL considered in <em>py-dsyre</em> as <strong>phenotype</strong>, that is, as possible models. With respect to Cartesian Genetic Programming, or other popular genetic programming approaches, <em>py-dsyre</em> does not define output nodes nor makes use of introns.</p>
<p>Lets see how this looks, first we import the module:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pydsyre as dsy
import numpy as np
</pre></div>
</div>
</div>
<p>We start defining the main object able to manipulate genotypes and phenotypes of our symbolic regression system, letting it know about the number of input terminals (variables and constants) and the kernels (binary and unary operations) we intend to use:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ex = dsy.expression(nvars=1,
                    ncons=1,
                    kernels=[&quot;sum&quot;, &quot;mul&quot;, &quot;div&quot;, &quot;sin&quot;])
</pre></div>
</div>
</div>
<p>we then define a genotype assembling triplets as detailed above:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>geno = [1,0,1,3,2,0,0,3,0,2,4,2]
</pre></div>
</div>
</div>
<p>Finally, we peek at the symbolic representation of the phenotype and see that indeed it contains the correct expressions:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(ex.sphenotype(geno=geno, vars=[&quot;x&quot;], cons=[&quot;c&quot;]))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;x&#39;, &#39;c&#39;, &#39;(x*c)&#39;, &#39;sin((x*c))&#39;, &#39;(sin((x*c))+x)&#39;, &#39;((sin((x*c))+x)/(x*c))&#39;]
</pre></div></div>
</div>
</section>
<section id="Genotype">
<h2>Genotype<a class="headerlink" href="#Genotype" title="Permalink to this heading">¶</a></h2>
<p>Now that we have understood how the basic idea works, we can quickly take a look at how to <strong>generate</strong> and <strong>manipulate</strong> genotypes. We start by creatin a random genotype of length 10. Since the expression has <em>nvars</em>=1 and <em>ncons</em>=1 the number of expressed models (dimensionality of the vector $[u_0, u_1, … ]) will be 12:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># We create a random genotype
geno = ex.random_genotype(length = 10)
# We compute the symbolic form of the phenotype
sphen = ex.sphenotype(geno=geno, vars=[&quot;x&quot;], cons=[&quot;c&quot;])
print(f&quot;Number of models: {len(sphen)}\nPhenotype: {sphen}&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of models: 12
Phenotype: [&#39;x&#39;, &#39;c&#39;, &#39;(c/c)&#39;, &#39;(c+(c/c))&#39;, &#39;(x+(c+(c/c)))&#39;, &#39;sin(x)&#39;, &#39;sin(x)&#39;, &#39;(sin(x)+c)&#39;, &#39;((c/c)+(x+(c+(c/c))))&#39;, &#39;(((c/c)+(x+(c+(c/c))))/(x+(c+(c/c))))&#39;, &#39;(sin(x)*(sin(x)+c))&#39;, &#39;(((c/c)+(x+(c+(c/c))))/(c/c))&#39;]
</pre></div></div>
</div>
<p>Let us now mutate at random three elements in the genotype and see what effect it has on the phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mutated_geno = ex.mutate(geno = geno, N = 3)
sphen = ex.sphenotype(geno=mutated_geno, vars=[&quot;x&quot;], cons=[&quot;c&quot;])
print(f&quot;Number of models: {len(sphen)}\nPhenotype: {sphen}&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of models: 12
Phenotype: [&#39;x&#39;, &#39;c&#39;, &#39;(c/c)&#39;, &#39;(c+(c/c))&#39;, &#39;(x+(c+(c/c)))&#39;, &#39;sin(x)&#39;, &#39;sin(x)&#39;, &#39;(sin(x)+c)&#39;, &#39;((c/c)+sin(x))&#39;, &#39;(((c/c)+sin(x))/(c+(c/c)))&#39;, &#39;(sin(x)*(sin(x)+c))&#39;, &#39;(((c/c)+sin(x))/(c/c))&#39;]
</pre></div></div>
</div>
</section>
<section id="Phenotype">
<h2>Phenotype<a class="headerlink" href="#Phenotype" title="Permalink to this heading">¶</a></h2>
<p>All the models contained in the phenotype need to be evaluated and assessed on data. To show how this is done we now make use of an expression with 10 variables and 3 constants:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ex = dsy.expression(nvars=3,
                    ncons=2,
                    kernels=[&quot;sum&quot;, &quot;mul&quot;, &quot;div&quot;, &quot;sin&quot;])
<br/></pre></div>
</div>
</div>
<p>and we generate a random genotype, this time of greater length:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>geno = ex.random_genotype(length = 100)
# Out of curiosity we also print the last expression in the phenotype
print(ex.sphenotype(geno)[-1])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(((((x2/x0)*x1)*x2)*(((x2/x0)*x1)*x2))+((sin(x0)*((x0/c0)/x0))*((((x2/x0)*x1)*x2)+c0)))
</pre></div></div>
</div>
<p>we create a meaningless dataset:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Data are just 1024 randomly distributed points
X = np.random.randn(1024, 3)
</pre></div>
</div>
</div>
<p>and compute the phenotype when the model parameters are (arbitrarily) [0.2, -0.4]:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>phen = ex.phenotype(geno = geno, xs = X, cons = [0.2, -0.4])
print(&quot;First 20 values in the phenotype for the first instance: &quot;, phen[0][:20])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
First 20 values in the phenotype for the first instance:  [-1.3137669274472548, -0.48195417432681165, 0.5837987234947484, 0.2, -0.4, -0.444370086731527, -0.9671493967307924, -6.568834637236273, 0.21416601824622677, 0.1250298480681002, 0.01563246290793222, 5.0, 0.21253257371430564, -0.9671493967307924, 0.19278166973072466, -0.26778815608058487, 0.11254569270198322, 0.3250298480681002, 0.9410322215777032, 0.21093616406148863]
</pre></div></div>
</div>
<p>and we time it:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%timeit ex.phenotype(geno = geno, xs = X, cons = [0.2, -0.4])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
32.5 µs ± 82.5 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
</pre></div></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">py-dsyre</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">A gentle introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Preamble">Preamble</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Genotype">Genotype</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Phenotype">Phenotype</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../theory/symbolic_regression.html">Symbolic Regression</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="../index.html" title="previous chapter">Welcome to py-dsyre’s documentation!</a></li>
      <li>Next: <a href="../theory/symbolic_regression.html" title="next chapter">Symbolic Regression</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, 2021, 2022 Francesco Biscani, Dario Izzo.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/notebooks/introduction.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>